src/magicbox/ResultPreviewsGrid.ts

```diff
@@ -0,0 +1,268 @@
+import { $$, Dom } from '../utils/Dom';
+import { defaults, findIndex } from 'underscore';
+import { Component } from '../ui/Base/Component';
+import { ISelectableItemsContainer } from './SuggestionsManager';
+
+export enum ResultPreviewsGridDirection {
+  Up = 'Up',
+  Down = 'Down',
+  Left = 'Left',
+  Right = 'Right'
+}
+
+export interface ISearchResultPreview {
+  element: HTMLElement;
+  onSelect?: () => void;
+}
+
+interface IActiveSearchResultPreview extends ISearchResultPreview {
+  deactivate: () => void;
+}
+
+export interface IResultPreviewsGridOptions {
+  selectedClass?: string;
+}
+
+export interface IPreviewHoveredEventArgs {
+  preview: ISearchResultPreview;
+}
+
+export enum ResultPreviewsGridEvents {
+  PreviewHovered = 'previewHovered'
+}
+
+export class ResultPreviewsGrid implements ISelectableItemsContainer<ISearchResultPreview, ResultPreviewsGridDirection> {
+  public static ContainerClassName = 'coveo-preview-container';
+  public static HeaderClassName = 'coveo-preview-header';
+  public static ResultsContainerClassName = 'coveo-preview-results';
+  public static PreviewIdPrefix = 'magic-box-preview-';
+
+  private static setItemIdOfElement(element: HTMLElement, id: number) {
+    element.id = this.PreviewIdPrefix + id.toString();
+  }
+
+  private static getItemIdFromElement(element: HTMLElement) {
+    const strId = element.id.substr(this.PreviewIdPrefix.length);
+    return strId ? parseInt(strId, 10) : null;
+  }
+
+  private root: HTMLElement;
+  private resultContainerElements?: {
+    container: Dom;
+    status: Dom;
+    results: Dom;
+  } = null;
+  private options: IResultPreviewsGridOptions;
+  private activePreviews: IActiveSearchResultPreview[] = [];
+  private keyboardSelectionMode: boolean = false;
+
+  constructor(private parentContainer: HTMLElement, options: IResultPreviewsGridOptions = {}) {
+    this.root = Component.resolveRoot(parentContainer);
+    this.options = defaults(options, <IResultPreviewsGridOptions>{
+      selectedClass: 'magic-box-selected'
+    });
+    this.buildResultsPreviewsContainer();
+  }
+
+  public isHoverKeyboardControlled() {
+    return this.keyboardSelectionMode;
+  }
+
+  public setDisplayedItems(items: ISearchResultPreview[]) {
+    this.clearDisplayedItems();
+    items.forEach(preview =>
+      this.appendSearchResultPreview({
+        element: preview.element.cloneNode(true) as HTMLElement,
+        onSelect: preview.onSelect
+      })
+    );
+  }
+
+  public getHoveredItem(): ISearchResultPreview {
+    if (this.activePreviews.length === 0) {
+      return null;
+    }
+    const previewId = this.getHoveredItemId();
+    return this.activePreviews[previewId] || null;
+  }
+
+  public tryHoverOnFirstItem() {
+    if (this.activePreviews.length === 0) {
+      return (this.keyboardSelectionMode = false);
```

Comment:
The boolean value was used to determine if moving the focus was successful or not, which would be used by the `SuggestionsManager` to determine if the hover should be moved back to the `SuggestionsList` component or not. However, this can already be detected by checking if getFocusedItem() was altered. I'll remove the return type and rename this method.